### PR TITLE
Added the compiler explanation to doc.md.

### LINKED ISSUE #180: [doc] The documentation for compiler
**Subject**
doc - compiler

**Detail**
Add the documentation for compiler to doc.md!

### PR DESCRIPTION
**Pull request by hihiboss**
resolved: #180 

details:

Documentation for the compiler!

~- [ ] Test case~

### CODE REVIEW

**Comment by hihiboss** [comment]
@zeroFruit @junbeomlee #181 
I had some problem, so I recreate the PR.

File: docs/doc.md [hunk 1]
```diff
@@ -134,6 +135,56 @@ That’s our lexer definition it has just token channel which are going to be us
 Syntax analysis reads the stream of `token` which is generated by `lexer`, then make AST(Abstract Synstax Tree) which is passed to compiler.  Koa parser is 'Pratt parser' which is easy to make, modulable and scalable. **The main idea of 'Pratt parser' is each `token` has its own parsing functions.** (infix parsing function, prefix parsing function). In the above diagram `Program` denotes root node of AST and AST consists of slice of `Statements`.
 
 
+### <a name="compile">Compile</a>
+
+Compiling produces a code by assembling information collected from other sources. Koa compiler **reads the `AST` made by `parser` and generates a code called `bytecode`.** `Bytecode` is a kind of assemble codes. This has an information how to execute a program.
+
+#### Bytecode
+
+`Bytecode` is an output generated by compiling. Through executing this code, a machine could execute a program.
```
Comment (zeroFruit) [inline docs/doc.md]: Suggestion!

`Bytecode` is an output generated by compiler. Through vm executing this code, we can get result of the source code

File: docs/doc.md [hunk 2]
```diff
@@ -134,6 +135,56 @@ That’s our lexer definition it has just token channel which are going to be us
...
+
+```go
+type Bytecode struct {
+   RawByte []byte
+   AsmCode []string
+   Abi     abi.Abi
+}
+```
+
+This is our `Bytecode` structure. It has 3 data. `RawByte` is a byte which has information to execute the program in hexadecimal. `AsmCode` is a collection of assemble codes more readable than byte to human. `Abi` is an interface needed to call the functions by user.
```
Comment (zeroFruit) [inline docs/doc.md]: suggestion!

This is our `Bytecode` structure. It has 3 fields. `RawByte` is a byte which has information of how to execute the program. And `RawByte` consists of hexadecimal code.

Comment (zeroFruit) [reply docs/doc.md]: suggestion!

`AsmCode` is a collection of assemble codes which is more readable to human because this is not bytes but string

Comment (zeroFruit) [reply docs/doc.md]: suggestion!

`Abi` is an interface needed to user for calling the functions.

File: docs/doc.md [hunk 3]
```diff
@@ -134,6 +135,56 @@ That’s our lexer definition it has just token channel which are going to be us
...
+
+[Function Jumper](#function-jumper) and [Function Selector](#function-selector) will be explained later.
+
+#### AST & Compile
+
+<p align="center"><img src="../image/ast-architecture.jpg" width="470px" height="340px"></p>
+
+The root of AST is a `contract`. `Contract` has some `functions`. And, a `function` has some `statements`.
+
+`CompileContract()` function compiles the contract. It compils every function and adds `VM memory settiong code` and `jump selector`. `compileFunction()` function compiles a function in the contract. It is called recursively and compiles each function. `compileStatement()` function compiles a statement in the function. It is called recursively and compiles each statement.
```
Comment (zeroFruit) [inline docs/doc.md]: typo! compils => compiles

Comment (zeroFruit) [reply docs/doc.md]: typo! settiong => setting

File: docs/doc.md [hunk 4]
```diff
@@ -134,6 +135,56 @@ That’s our lexer definition it has just token channel which are going to be us
...
+
+#### <a name="function-selector">Function Selector</a>
+
+Each function should have identifiers to distinguish each other. The `function selector` is an identifier to distinguish each function. It is generated by Keccak algorithm (SHA-3). This algorithm generates a bytes with function name and paramenter type. And the first 4 bytes of those bytes are the `function selector`. For example, the `function selector` of `function foo(int a) bool` is `0x4ff9f498`. The keccak algorithm receives `"foo(int)"` and generates some bytes. The first 4 bytes of those are `0x4ff9f598` and this is the `function selector`.
```
Comment (zeroFruit) [inline docs/doc.md]: typo! paramenter => parameter

**Review by zeroFruit — APPROVED** [review]
Good job! 👍

File: docs/doc.md [hunk 5]
```diff
@@ -134,6 +135,56 @@ That’s our lexer definition it has just token channel which are going to be us
...
+
+```go
+type Bytecode struct {
+   RawByte []byte
+   AsmCode []string
+   Abi     abi.Abi
+}
+```
+
+This is our `Bytecode` structure. It has 3 fields. `RawByte` is a byte which has information of how to execute the program. And `RawByte` consists of hexadecimal code. `AsmCode` is a collection of assemble codes which is more readable to human than bytes. `Abi` is an interface needed to user for calling the functions.
```
Comment (junbeomlee) [inline docs/doc.md]: `RawByte` is a byte which has information of how to execute the program.
`RawByte` is s program!! not an information of how to execute the program.
plz fix the sentence!

File: docs/doc.md [hunk 6]
```diff
@@ -134,6 +135,56 @@ That’s our lexer definition it has just token channel which are going to be us
...
+
+#### <a name="function-selector">Function Selector</a>
+
+Each function should have identifiers to distinguish each other. The `function selector` is an identifier to distinguish each function. It is generated by Keccak algorithm (SHA-3). This algorithm generates a bytes with function name and parameter type. And the first 4 bytes of those bytes are the `function selector`. For example, the `function selector` of `function foo(int a) bool` is `0x4ff9f498`. The keccak algorithm receives `"foo(int)"` and generates some bytes. The first 4 bytes of those are `0x4ff9f598` and this is the `function selector`.
+
+#### <a name="function-jumper">Function Jumper</a>
+
+The `function jumper` includes a logic to find a function to call. If an user calls `function a()`, program counter points `function jumper`. Because of the static size of `function jumper`, the program could know the position of `function jumper`. This finds the `function selector` corresponding with `function a()`. Then, returns the bytes of `function a()`'s `function selector`. Finally, `VM` could know the position `function a()` and execute it.
```
Comment (junbeomlee) [inline docs/doc.md]: Not because of the static size of `function jumper`, but because of the static location of `function jumper`

**Comment by junbeomlee** [comment]
Except the two comments!! nice work!

**Pull request merged**